lib/HttpClient.ts

```diff
@@ -226,63 +247,93 @@ export class HttpClient {
         this._disposed = true;
     }
 
-    private _requestRaw(info: RequestInfo, data: string | NodeJS.ReadableStream): Promise<HttpClientResponse> {
-        return new Promise<HttpClientResponse>((resolve, reject) => {
-            let socket;
+    /**
+     * Raw request.
+     * @param info 
+     * @param data 
+     */
+    public requestRaw(info: ifm.IRequestInfo, data: string | NodeJS.ReadableStream): Promise<ifm.IHttpClientResponse> {
+        return new Promise<ifm.IHttpClientResponse>((resolve, reject) => {
+            var callbackForResult = function (err: any, res: ifm.IHttpClientResponse) {
+                if (err) {
+                    reject(err);
+                }
 
-            let isDataString = typeof (data) === 'string';
+                resolve(res);
+            };
 
-            if (typeof (data) === 'string') {
-                info.options.headers["Content-Length"] = Buffer.byteLength(data, 'utf8');
-            }
+            this.requestRawWithCallback(info, data, callbackForResult);
+        });
+    }
 
-            let req: http.ClientRequest = info.httpModule.request(info.options, (msg: http.IncomingMessage) => {
-                let res: HttpClientResponse = new HttpClientResponse(msg);
-                resolve(res);
-            });
+    /**
+     * Raw request with callback.
+     * @param info 
+     * @param data 
+     * @param onResult 
+     */
+    public requestRawWithCallback(info: ifm.IRequestInfo, data: string | NodeJS.ReadableStream, onResult: (err: any, res: ifm.IHttpClientResponse) => void): void {
```

Comment:
have you considered an actual overload rather than two functions `requestRaw` and `requestRawWithCallback`?

Reply:
or is the return type different, is that the point?

Reply:
do these both need to be public? do we want to support both interfaces?

Reply:
(also, if i remember correctly, i think overloads can have different return types so even if we want to support both as public i think it could simply be an overload. i think that would be a more common API interface, unless these function names were chosen to match some other library's pattern (e.g. node)?

Reply:
I tried to do an overload and the compiler was not happy with it, I can look into it more deeply. We may not need the non-callback version to be public but we need the callback version to be public for the authorization workflow. It has an HttpClient that it needs to call this method on in order to make the NTLM requests.

Yes return type and params are different, one returns the response and takes requestinfo and data and one is void but takes in a callback and input is requestinfo, data, and a callback. I looked up a bit on the overloading and it seems like you can only declare the overload but not implement it, there can only be one implementation. For instance add(string foo); add(int foo) {}. But only the second has an implementation and you check typeof foo. I could be misunderstanding it.